upstairs/src/dummy_downstairs_tests.rs

```diff
@@ -2380,75 +2122,30 @@ async fn test_deactivate_slow() {
-        // response to come back before returning
-        tokio::spawn(async move {
-            let mut buffer = Buffer::new(1, 512);
-            guest.read(Block::new_512(0), &mut buffer).await.unwrap();
-        });
-    }
+    // We must `spawn` here because `read` will wait for the
+    // response to come back before returning
+    harness.spawn(|guest| async move {
+        let mut buffer = Buffer::new(1, 512);
+        guest.read(Block::new_512(0), &mut buffer).await.unwrap();
+    });
 
     // Ensure that all three clients got the read request
-    let job_id = match harness.ds1().recv().await.unwrap() {
-        Message::ReadRequest { job_id, .. } => job_id,
-        _ => panic!("invalid request"),
-    };
-    assert!(matches!(
-        harness.ds2.recv().await.unwrap(),
-        Message::ReadRequest { .. },
-    ));
-    assert!(matches!(
-        harness.ds3.recv().await.unwrap(),
-        Message::ReadRequest { .. },
-    ));
+    harness.ds1().ack_read().await;
+    harness.ds2.ack_read().await;
+    harness.ds3.ack_read().await;
 
-    let (block, data) = make_blank_read_response();
-    let response = Message::ReadResponse {
-        header: ReadResponseHeader {
-            upstairs_id: harness.guest.get_uuid().await.unwrap(),
-            session_id: harness.ds1().upstairs_session_id.unwrap(),
-            job_id,
-            blocks: Ok(vec![block.clone()]),
-        },
-        data,
-    };
-    harness.ds1().send(response.clone()).unwrap();
-    harness.ds2.send(response.clone()).unwrap();
-    harness.ds3.send(response.clone()).unwrap();
-
-    let deactivate_handle = {
-        // Send a deactivate request.
-        let guest = harness.guest.clone();
-        tokio::spawn(async move { guest.deactivate().await })
-    };
-
-    let job_id = match harness.ds1().recv().await.unwrap() {
-        Message::Flush { job_id, .. } => job_id,
-        _ => panic!("invalid request"),
-    };
-    assert!(matches!(
-        harness.ds2.recv().await.unwrap(),
-        Message::Flush { .. },
-    ));
-    assert!(matches!(
-        harness.ds3.recv().await.unwrap(),
-        Message::Flush { .. },
-    ));
+    // Send a deactivate request.  This sends a final flush
+    let deactivate_handle =
+        harness.spawn(|guest| async move { guest.deactivate().await });
 
     // Reply on ds1, which concludes the deactivation of this downstairs
-    let response = Message::FlushAck {
-        upstairs_id: harness.guest.get_uuid().await.unwrap(),
-        session_id: harness.ds1().upstairs_session_id.unwrap(),
-        job_id,
-        result: Ok(()),
-    };
-    harness.ds1().send(response.clone()).unwrap();
+    harness.ds1().ack_flush().await;
 
     // At this point, the upstairs sends YouAreNoLongerActive, but then
     // drops the sender end, so we can't actually see it.
+    //
+    // TODO is this correct?  Shouldn't values in the mpsc queue be read out,
+    // even if the sender has been dropped?
```

Comment:
I don't think our `ClientIoTask` drains the requests before stopping: a quick glace at client.rs and it looks like landing in a `self.stop` branch breaks out of the task?

Reply:
That makes sense, I forgot that the `ClientIoTask` drops values while exiting.

Reply:
(removed the comment in a64ce0c5124e7668d876634fbc2b2d675665fbcd)